external-crates/move/crates/move-compiler/src/parser/ast.rs

```diff
@@ -301,16 +306,65 @@ pub struct Constant {
 // Types
 //**************************************************************************************************
 
+// MName = Name <TyArgs>
+//       | Leading <TyArgs> :: Name
+//       | Leading :: Name <TyArgs>
+//       | Leading :: Name <TyArgs> :: Name
+//       | Leading :: Name :: Name <TyArgs>
+//       | Leading :: Name :: Name <TyArgs> :: Name
+//
+// VName = MName <TyArgs> :: Name
+
 // A ModuleAccess references a local or global name or something from a module,
 // either a struct type or a function.
+// The boolean flag indicates if a name was pushed on after tyarg parsing, which we use in
+// expansion to determine if the variant form was valid.
+// FIXME(cswords): We should eventually rewrite all of this to be `NamePath`s that hold tyargs:
```

Comment:
I don't quite understand the FIXME, in what you were going for below

Reply:
Old comment from before I did the work "properly" -- I was going to do the ad-hoc solution we suggested. Sorry for the confusion.